lib/collection/src/discovery.rs

```diff
@@ -0,0 +1,234 @@
+    let request_batch = DiscoverRequestBatch {
+        searches: vec![request],
+    };
+    let results = discover_batch(
+        request_batch,
+        collection,
+        collection_by_name,
+        read_consistency,
+    )
+    .await?;
+    Ok(results.into_iter().next().unwrap())
+}
+
+pub async fn discover_batch<'a, F, Fut>(
+    request_batch: DiscoverRequestBatch,
+    collection: &Collection,
+    collection_by_name: F,
+    read_consistency: Option<ReadConsistency>,
+) -> CollectionResult<Vec<Vec<ScoredPoint>>>
+where
+    F: Fn(String) -> Fut,
+    Fut: Future<Output = Option<RwLockReadGuard<'a, Collection>>>,
+{
+    // shortcuts batch if all requests with limit=0
+    if request_batch.searches.iter().all(|s| s.limit == 0) {
+        return Ok(vec![]);
+    }
+
+    // Validate context_pairs and/or target have value(s)
+    request_batch.searches.iter().try_for_each(|request| {
+        let no_pairs = request.context_pairs.is_none()
+            || request
+                .context_pairs
+                .as_ref()
+                .is_some_and(|pairs| pairs.is_empty());
+
+        let no_target = request.target.is_none();
+
+        if no_pairs && no_target {
+            return Err(CollectionError::bad_request(
+                "target and/or context_pairs must be specified".to_string(),
+            ));
+        }
+
+        Ok(())
+    })?;
+
+    // Pack all referenced ids for retrieving them once
+    let mut id_references = ReferencedPoints::new();
+    request_batch.searches.iter().for_each(|request| {
+        let collection_name = request.lookup_from.as_ref().map(|s| &s.collection);
+
+        let vector_name = get_vector_name(request);
+
+        let ids_iter = iterate_examples(request).filter_map(RecommendExample::as_point_id);
+
+        id_references.add_from_iter(ids_iter, vector_name, collection_name);
+    });
+
+    // Fetch all referenced vectors
+    let referenced_vectors = id_references
+        .fetch_vectors(collection, read_consistency, collection_by_name)
+        .await?;
+
+    // Create core search requests
+    let mut core_searches = Vec::with_capacity(request_batch.searches.len());
+    for request in &request_batch.searches {
+        let lookup_collection_name = request.lookup_from.as_ref().map(|x| &x.collection);
+
+        let lookup_vector_name = get_vector_name(request);
+
+        // Check we actually fetched all referenced vectors in this request
+        let referenced_ids = iterate_examples(request)
+            .filter_map(RecommendExample::as_point_id)
+            .collect_vec();
+
+        referenced_ids.iter().try_for_each(|&point_id| {
+            if !referenced_vectors.contains_key(&PointRef {
+                collection_name: lookup_collection_name,
+                point_id,
+            }) {
+                return Err(CollectionError::PointNotFound {
+                    missed_point_id: point_id,
+                });
+            }
+            Ok(())
+        })?;
```

Comment:
A `.iter().find()` or `if .iter().any()` may be a bit nicer here. What do you think?